src/cast-dispatch.c

```diff
@@ -16,116 +9,39 @@ SEXP vec_cast_dispatch(SEXP x,
-    case vctrs_class_bare_factor:
-      return fct_as_character(x, x_arg);
-
-    case vctrs_class_bare_ordered:
-      return ord_as_character(x, x_arg);
-
-    default:
-      break;
-    }
-    break;
-
-  case vctrs_type_dataframe:
-    switch (class_type(x)) {
-    case vctrs_class_bare_data_frame:
-      Rf_errorcall(R_NilValue, "Internal error: `x` should have been classified as a `vctrs_type_dataframe`");
-
-    case vctrs_class_bare_tibble:
-      return df_cast(x, to, x_arg, to_arg);
-
-    default:
-      break;
-    }
-    break;
-
-  case vctrs_type_s3:
-    return vec_cast_dispatch2(x, to, x_type, lossy, x_arg, to_arg);
+  int dir = 0;
+  enum vctrs_type2_s3 type2_s3 = vec_typeof2_s3_impl(x, to, x_type, to_type, &dir);
 
-  default:
-    break;
-  }
-
-  return R_NilValue;
-}
+  switch (type2_s3) {
 
-
-static SEXP vec_cast_dispatch2(SEXP x, SEXP to,
-                               enum vctrs_type x_type,
-                               bool* lossy,
-                               struct vctrs_arg* x_arg,
-                               struct vctrs_arg* to_arg) {
-  switch (class_type(to)) {
-  case vctrs_class_bare_factor:
-    switch (x_type) {
-    case vctrs_type_character:
+  case vctrs_type2_s3_character_bare_factor:
+    if (dir == 0) {
       return chr_as_factor(x, to, lossy, to_arg);
-
-    case vctrs_type_s3:
-      switch (class_type(x)) {
-      case vctrs_class_bare_factor:
-        return fct_as_factor(x, to, lossy, x_arg, to_arg);
-
-      default:
-        break;
-      }
-
-    default:
-      break;
+    } else {
+      return fct_as_character(x, x_arg);
     }
-    break;
 
-  case vctrs_class_bare_ordered:
-    switch (x_type) {
-    case vctrs_type_character:
+  case vctrs_type2_s3_character_bare_ordered:
+    if (dir == 0) {
       return chr_as_ordered(x, to, lossy, to_arg);
-
-    case vctrs_type_s3:
-      switch (class_type(x)) {
-      case vctrs_class_bare_ordered:
-        return ord_as_ordered(x, to, lossy, x_arg, to_arg);
-
-      default:
-        break;
-      }
-
-    default:
-      break;
+    } else {
+      return ord_as_character(x, x_arg);
     }
-    break;
-
-  case vctrs_class_bare_tibble:
-    switch (x_type) {
-    case vctrs_type_dataframe:
-      return df_cast(x, to, x_arg, to_arg);
-
-    case vctrs_type_s3:
-      switch (class_type(x)) {
-      case vctrs_class_bare_data_frame:
-        Rf_errorcall(R_NilValue, "Internal error: `x` should have been classified as a `vctrs_type_dataframe`");
 
-      case vctrs_class_bare_tibble:
-        return df_cast(x, to, x_arg, to_arg);
+  case vctrs_type2_s3_bare_factor_bare_factor:
+    return fct_as_factor(x, to, lossy, x_arg, to_arg);
 
-      default:
-        break;
-      }
+  case vctrs_type2_s3_bare_ordered_bare_ordered:
+    return ord_as_ordered(x, to, lossy, x_arg, to_arg);
 
-    default:
-      break;
+  case vctrs_type2_s3_dataframe_bare_tibble:
+    if (dir == 0) {
+      return tib_cast(x, to, x_arg, to_arg);
```

Comment:
We'll probably need to export these methods publicly so that cast-methods can call them, since we're going to remove inheritance and `NextMethod()` (which is buggy right now with double dispatch). I think I prefer this naming scheme because it seems simpler. Also we need to consider how the exported `cast` methods integrate with the `as` conversion methods. We can get a feel for it, and revisit when it's time to export?

Reply:
Something about the name is confusing me. Is it backwards? Don't we normally make the output the suffix? So: `cast_tib()` and `cast_df()`? I guess it is somewhat ambiguous because `to` is also an input

Reply:
In that case it would feel more consistent with `new_data_frame()` and `as_data_frame()` if they were named `vec_cast_tibble()` / `vec_cast_data_frame()`?

hmm let's keep thinking about it.

Reply:
If we export these like this, would probably need to rename `vec_list_cast()` to `vec_cast_list()` as well

Reply:
I think `vec_list_cast()` will be deprecated in favour of `vec_chop()` and `vec_chop2()`.